File: src/utils/links.rs
//! Link management utilities
//! 
//! This module provides utilities for finding, validating, and manipulating
//! links within notes and between notes.

#![allow(dead_code)]

use crate::models::{ParsedLink, LinkType, LinkReference};
use crate::error::{Result, TriliumError};
use regex::Regex;
use std::collections::HashMap;
use std::sync::OnceLock;

/// Static regex for link parsing to avoid recompilation
static LINK_REGEX: OnceLock<Regex> = OnceLock::new();

/// Get or create the link regex pattern
fn get_link_regex() -> &'static Regex {
    LINK_REGEX.get_or_init(|| {
        Regex::new(r"\[\[([^\]|]+)(?:\|([^\]]+))?\]\]").unwrap()
    })
}

/// Precomputed line offset cache for efficient line number lookups
#[derive(Debug, Clone)]
struct LineOffsetCache {
    line_starts: Vec<usize>,
    content_length: usize,
}

impl LineOffsetCache {
    fn new(content: &str) -> Self {
        let mut line_starts = vec![0];
        
        // Pre-compute all line start positions
        for (pos, byte) in content.bytes().enumerate() {
            if byte == b'\n' {
                line_starts.push(pos + 1);
            }
        }
        
        Self {
            line_starts,
            content_length: content.len(),
        }
    }
    
    /// Find line number for a given byte position using binary search
    fn find_line_number(&self, pos: usize) -> usize {
        if pos >= self.content_length {
            return self.line_starts.len().saturating_sub(1);
        }
        
        // Binary search for efficiency
        match self.line_starts.binary_search(&pos) {
            Ok(line) => line,
            Err(line) => line.saturating_sub(1),
        }
    }
    
    /// Get the content of a specific line
    fn get_line_content<'a>(&self, content: &'a str, line_num: usize) -> &'a str {
        if line_num >= self.line_starts.len() {
            return "";
        }
        
        let start = self.line_starts[line_num];
        let end = if line_num + 1 < self.line_starts.len() {
            self.line_starts[line_num + 1].saturating_sub(1) // Exclude newline
        } else {
            self.content_length
        };
        
        &content[start..end.min(self.content_length)]
    }
}

/// Parse wiki-style links from note content with optimized performance
/// Supports formats: [[note-id]], [[Note Title]], [[note-id|Custom Text]], [[Note Title|Custom Text]]
pub fn parse_links(content: &str) -> Result<Vec<ParsedLink>> {
    // Security: Validate content size to prevent memory exhaustion
    const MAX_CONTENT_SIZE: usize = 10_000_000; // 10MB limit
    if content.len() > MAX_CONTENT_SIZE {
        return Err(TriliumError::SecurityError(
            format!("Content too large for link parsing (max {} bytes)", MAX_CONTENT_SIZE)
        ));
    }
    
    let mut links = Vec::new();
    links.reserve(100); // Pre-allocate reasonable capacity
    
    // Use cached regex for better performance
    let link_regex = get_link_regex();
    
    // Security: Limit number of links to prevent DoS
    const MAX_LINKS: usize = 1000;
    let mut link_count = 0;
    
    for cap in link_regex.captures_iter(content) {
        if link_count >= MAX_LINKS {
            eprintln!("Warning: Too many links found, limiting to {}", MAX_LINKS);
            break;
        }
        
        let full_match = cap.get(0).unwrap();
        let target = cap.get(1).unwrap().as_str().trim();
        let display_text = cap.get(2).map(|m| m.as_str().trim().to_string());
        
        // Security: Validate link target length
        const MAX_TARGET_LENGTH: usize = 500;
        if target.len() > MAX_TARGET_LENGTH {
            continue; // Skip oversized targets
        }
        
        // Security: Validate display text length
        if let Some(ref text) = display_text {
            const MAX_DISPLAY_TEXT_LENGTH: usize = 500;
            if text.len() > MAX_DISPLAY_TEXT_LENGTH {
                continue; // Skip oversized display text
            }
        }
        
        // Determine if target is likely an ID (contains numbers/hyphens) or a title
        let link_type = if is_likely_note_id(target) {
            LinkType::NoteId
        } else {
            LinkType::NoteTitle
        };
        
        links.push(ParsedLink {
            link_type,
            target: target.to_string(),
            display_text,
            start_pos: full_match.start(),
            end_pos: full_match.end(),
        });
        
        link_count += 1;
    }
    
    Ok(links)
}

/// Check if a string looks like a Trilium note ID
fn is_likely_note_id(s: &str) -> bool {
    // Trilium note IDs are typically alphanumeric strings
    s.len() >= 10 && s.chars().all(|c| c.is_ascii_alphanumeric())
}

/// Extract link references with context from content using optimized line lookup
pub fn extract_link_references(
    content: &str, 
    from_note_id: &str, 
    from_title: &str
) -> Result<Vec<LinkReference>> {
    // Security: Validate input sizes
    const MAX_ID_LENGTH: usize = 100;
    const MAX_TITLE_LENGTH: usize = 500;
    
    if from_note_id.len() > MAX_ID_LENGTH {
        return Err(TriliumError::ValidationError(
            "from_note_id too long".to_string()
        ));
    }
    if from_title.len() > MAX_TITLE_LENGTH {
        return Err(TriliumError::ValidationError(
            "from_title too long".to_string()
        ));
    }
    
    let links = parse_links(content)?;
    
    // Pre-compute line offsets for efficient line number lookups
    let line_cache = LineOffsetCache::new(content);
    
    let mut references = Vec::new();
    references.reserve(links.len()); // Pre-allocate for efficiency
    
    for link in links {
        // Use optimized line lookup
        let line_num = line_cache.find_line_number(link.start_pos);
        let context = line_cache.get_line_content(content, line_num).to_string();
        
        let link_text = link.display_text.unwrap_or_else(|| link.target.clone());
        
        references.push(LinkReference {
            from_note_id: from_note_id.to_string(),
            to_note_id: link.target.clone(),
            from_title: from_title.to_string(),
            link_text,
            context,
        });
    }
    
    Ok(references)
}

/// Replace or update links in content with validation
pub fn replace_links(content: &str, replacements: &HashMap<String, String>) -> Result<String> {
    // Security: Validate content size
    const MAX_CONTENT_SIZE: usize = 10_000_000; // 10MB limit
    if content.len() > MAX_CONTENT_SIZE {
        return Err(TriliumError::SecurityError(
            format!("Content too large for link replacement (max {} bytes)", MAX_CONTENT_SIZE)
        ));
    }
    
    // Security: Validate replacements map size
    const MAX_REPLACEMENTS: usize = 10_000;
    if replacements.len() > MAX_REPLACEMENTS {
        return Err(TriliumError::SecurityError(
            format!("Too many replacements (max {})", MAX_REPLACEMENTS)
        ));
    }
    
    // Use cached regex
    let link_regex = get_link_regex();
    
    let result = link_regex.replace_all(content, |caps: &regex::Captures| {
        let target = caps.get(1).unwrap().as_str().trim();
        let display_text = caps.get(2).map(|m| m.as_str().trim());
        
        if let Some(new_target) = replacements.get(target) {
            // Security: Validate replacement target length
            if new_target.len() > 500 {
                return caps.get(0).unwrap().as_str().to_string(); // Skip replacement if too long
            }
            
            if let Some(display) = display_text {
                format!("[[{}|{}]]", new_target, display)
            } else {
                format!("[[{}]]", new_target)
            }
        } else {
            caps.get(0).unwrap().as_str().to_string()
        }
    });
    
    Ok(result.to_string())
}

/// Generate markdown-style links from wiki links for export with validation
pub fn convert_to_markdown_links(content: &str, note_titles: &HashMap<String, String>) -> Result<String> {
    // Security: Validate content size
    const MAX_CONTENT_SIZE: usize = 10_000_000; // 10MB limit
    if content.len() > MAX_CONTENT_SIZE {
        return Err(TriliumError::SecurityError(
            format!("Content too large for markdown conversion (max {} bytes)", MAX_CONTENT_SIZE)
        ));
    }
    
    // Security: Validate note titles map size
    const MAX_TITLES: usize = 10_000;
    if note_titles.len() > MAX_TITLES {
        return Err(TriliumError::SecurityError(
            format!("Too many note titles (max {})", MAX_TITLES)
        ));
    }
    
    // Use cached regex
    let link_regex = get_link_regex();
    
    let result = link_regex.replace_all(content, |caps: &regex::Captures| {
        let target = caps.get(1).unwrap().as_str().trim();
        let display_text = caps.get(2).map(|m| m.as_str().trim());
        
        let link_text = display_text.unwrap_or(
            note_titles.get(target).map(|s| s.as_str()).unwrap_or(target)
        );
        
        // Security: Validate lengths to prevent output bloat
        if link_text.len() > 200 || target.len() > 100 {
            return caps.get(0).unwrap().as_str().to_string(); // Keep original if too long
        }
        
        format!("[{}](trilium://note/{})", link_text, target)
    });
    
    Ok(result.to_string())
}

/// Find broken links (links that point to non-existent notes)
pub fn find_broken_links<'a>(links: &'a [ParsedLink], existing_note_ids: &[String]) -> Vec<&'a ParsedLink> {
    let existing_set: std::collections::HashSet<&String> = existing_note_ids.iter().collect();
    
    links.iter()
        .filter(|link| match link.link_type {
            LinkType::NoteId => !existing_set.contains(&link.target),
            LinkType::NoteTitle => false, // Title-based links are harder to validate without full lookup
        })
        .collect()
}

/// Auto-complete suggestions for link typing
pub fn suggest_link_completions(
    input: &str, 
    available_notes: &[(String, String)] // (id, title) pairs
) -> Vec<(String, String)> {
    let input_lower = input.to_lowercase();
    let mut suggestions = Vec::new();
    
    for (id, title) in available_notes {
        // Match by title (partial, case-insensitive)
        if title.to_lowercase().contains(&input_lower) {
            suggestions.push((id.clone(), title.clone()));
        }
        // Match by ID prefix
        else if id.to_lowercase().starts_with(&input_lower) {
            suggestions.push((id.clone(), title.clone()));
        }
    }
    
    // Sort by relevance (exact matches first, then prefix matches, then contains)
    suggestions.sort_by(|a, b| {
        let a_title_lower = a.1.to_lowercase();
        let b_title_lower = b.1.to_lowercase();
        
        let a_score = if a_title_lower == input_lower { 0 }
                     else if a_title_lower.starts_with(&input_lower) { 1 }
                     else { 2 };
        
        let b_score = if b_title_lower == input_lower { 0 }
                     else if b_title_lower.starts_with(&input_lower) { 1 }
                     else { 2 };
        
        a_score.cmp(&b_score).then_with(|| a.1.cmp(&b.1))
    });
    
    suggestions.truncate(10); // Limit to top 10 suggestions
    suggestions
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_simple_links() {
        let content = "See [[important-note]] and [[Another Note]] for details.";
        let links = parse_links(content).unwrap();
        
        assert_eq!(links.len(), 2);
        assert_eq!(links[0].target, "important-note");
        assert_eq!(links[0].link_type, LinkType::NoteId);
        assert_eq!(links[1].target, "Another Note");
        assert_eq!(links[1].link_type, LinkType::NoteTitle);
    }

    #[test]
    fn test_parse_links_with_display_text() {
        let content = "Check [[note123|Custom Label]] and [[Long Note Title|Short]]";
        let links = parse_links(content).unwrap();
        
        assert_eq!(links.len(), 2);
        assert_eq!(links[0].target, "note123");
        assert_eq!(links[0].display_text, Some("Custom Label".to_string()));
        assert_eq!(links[1].target, "Long Note Title");
        assert_eq!(links[1].display_text, Some("Short".to_string()));
    }

    #[test]
    fn test_is_likely_note_id() {
        assert!(is_likely_note_id("abc123def456"));
        assert!(is_likely_note_id("1234567890"));
        assert!(!is_likely_note_id("Note Title"));
        assert!(!is_likely_note_id("short"));
        assert!(!is_likely_note_id("has spaces"));
    }

    #[test]
    fn test_replace_links() {
        let content = "See [[old-id]] and [[old-title|Display]] for info.";
        let mut replacements = HashMap::new();
        replacements.insert("old-id".to_string(), "new-id".to_string());
        replacements.insert("old-title".to_string(), "new-title".to_string());
        
        let result = replace_links(content, &replacements).unwrap();
        assert_eq!(result, "See [[new-id]] and [[new-title|Display]] for info.");
    }
    
    #[test]
    fn test_line_offset_cache() {
        let content = "Line 1\nLine 2\nLine 3\nLine 4";
        let cache = LineOffsetCache::new(content);
        
        // Test line number finding
        assert_eq!(cache.find_line_number(0), 0); // Start of line 1
        assert_eq!(cache.find_line_number(7), 1); // Start of line 2
        assert_eq!(cache.find_line_number(14), 2); // Start of line 3
        
        // Test line content extraction
        assert_eq!(cache.get_line_content(content, 0), "Line 1");
        assert_eq!(cache.get_line_content(content, 1), "Line 2");
        assert_eq!(cache.get_line_content(content, 2), "Line 3");
        assert_eq!(cache.get_line_content(content, 3), "Line 4");
    }
    
    #[test]
    fn test_extract_link_references_optimized() {
        let content = "First line\nSee [[test-note]] for details\nLast line";
        let references = extract_link_references(content, "source123", "Source Note").unwrap();
        
        assert_eq!(references.len(), 1);
        assert_eq!(references[0].to_note_id, "test-note");
        assert_eq!(references[0].context, "See [[test-note]] for details");
        assert_eq!(references[0].from_note_id, "source123");
    }
    
    #[test]
    fn test_security_limits() {
        // Test content size limit for parsing
        let large_content = "a".repeat(15_000_000); // 15MB
        let result = parse_links(&large_content);
        assert!(result.is_err());
        
        // Test too many replacements
        let content = "[[test]]";
        let mut replacements = HashMap::new();
        for i in 0..15_000 {
            replacements.insert(format!("key{}", i), format!("value{}", i));
        }
        let result = replace_links(content, &replacements);
        assert!(result.is_err());
    }
    
    #[test]
    fn test_performance_with_many_links() {
        // Test performance with many links
        let mut content = String::new();
        for i in 0..500 {
            content.push_str(&format!("Link to [[note{}]] here. ", i));
        }
        
        let start = std::time::Instant::now();
        let links = parse_links(&content).unwrap();
        let duration = start.elapsed();
        
        assert_eq!(links.len(), 500);
        assert!(duration.as_millis() < 100, "Link parsing took too long: {:?}", duration);
    }
}

File: src/utils/links.ts
/**
 * Link Management Utilities
 * 
 * Provides comprehensive wiki-style link parsing, validation, and management
 * with performance optimizations and backlink tracking.
 */

import type { EntityId, ParsedLink, LinkReference, Note, NoteWithContent } from '../types/api.js';
import { LinkType } from '../types/api.js';

/**
 * Wiki-style link patterns with optimized regex
 */
const LINK_PATTERNS = {
  // [[noteId]] or [[noteId|display text]]
  NOTE_ID: /\[\[([a-zA-Z0-9_-]{10,22})(?:\|([^\]]+))?\]\]/g,
  
  // [[Note Title]] or [[Note Title|display text]]
  NOTE_TITLE: /\[\[([^|\]]+?)(?:\|([^\]]+))?\]\]/g,
  
  // Combined pattern for efficient single-pass parsing
  COMBINED: /\[\[([^|\]]+?)(?:\|([^\]]+))?\]\]/g
} as const;

/**
 * Link validation patterns
 */
const VALIDATION_PATTERNS = {
  // Valid note ID format (10-22 alphanumeric characters with dashes/underscores)
  NOTE_ID: /^[a-zA-Z0-9_-]{10,22}$/,
  
  // Valid title format (non-empty, reasonable length)
  NOTE_TITLE: /^.{1,500}$/,
  
  // Invalid characters for display text
  INVALID_DISPLAY: /[\[\]]/
} as const;

/**
 * Performance optimization: Line offset cache for large documents
 */
interface LineOffsetCache {
  offsets: number[];
  lastUpdate: number;
  content: string;
}

const lineOffsetCaches = new Map<string, LineOffsetCache>();

/**
 * Cache management configuration
 */
const CACHE_CONFIG = {
  MAX_CACHE_SIZE: 100,
  CACHE_TTL_MS: 5 * 60 * 1000, // 5 minutes
  CLEANUP_INTERVAL_MS: 60 * 1000 // 1 minute
} as const;

/**
 * Parse wiki-style links from content with performance optimizations
 */
export function parseLinks(content: string, noteId?: EntityId): ParsedLink[] {
  if (!content || content.length === 0) {
    return [];
  }

  const links: ParsedLink[] = [];
  const seenLinks = new Set<string>(); // Deduplicate links
  
  // Reset regex state
  LINK_PATTERNS.COMBINED.lastIndex = 0;
  
  let match: RegExpExecArray | null;
  
  while ((match = LINK_PATTERNS.COMBINED.exec(content)) !== null) {
    const [fullMatch, target, displayText] = match;
    if (!target) continue;
    
    const startPos = match.index;
    const endPos = startPos + fullMatch.length;
    
    // Determine link type
    const linkType = VALIDATION_PATTERNS.NOTE_ID.test(target) 
      ? LinkType.NoteId 
      : LinkType.NoteTitle;
    
    // Create unique key for deduplication
    const linkKey = `${linkType}:${target}:${displayText || ''}`;
    
    if (seenLinks.has(linkKey)) {
      continue;
    }
    seenLinks.add(linkKey);
    
    // Validate link
    if (!validateLinkTarget(target, linkType)) {
      continue;
    }
    
    if (displayText && !validateDisplayText(displayText)) {
      continue;
    }
    
    links.push({
      linkType,
      target: target.trim(),
      displayText: displayText?.trim(),
      startPos,
      endPos
    });
  }
  
  return links;
}

/**
 * Parse links with line number information for better context
 */
export function parseLinksWithLineNumbers(content: string, noteId?: EntityId): Array<ParsedLink & { lineNumber: number; columnNumber: number }> {
  const links = parseLinks(content, noteId);
  const lineOffsets = getLineOffsets(content, noteId);
  
  return links.map(link => {
    const { lineNumber, columnNumber } = getLineAndColumn(link.startPos, lineOffsets);
    
    return {
      ...link,
      lineNumber,
      columnNumber
    };
  });
}

/**
 * Extract outgoing links from note content
 */
export function getOutgoingLinks(note: NoteWithContent): ParsedLink[] {
  if (!note.content) {
    return [];
  }
  
  return parseLinks(note.content, note.noteId);
}

/**
 * Find backlinks to a specific note across a collection of notes
 */
export function findBacklinks(
  targetNoteId: EntityId,
  targetTitle: string,
  notes: NoteWithContent[]
): LinkReference[] {
  const backlinks: LinkReference[] = [];
  
  for (const note of notes) {
    if (!note.content || note.noteId === targetNoteId) {
      continue;
    }
    
    const links = parseLinks(note.content, note.noteId);
    
    for (const link of links) {
      let isMatch = false;
      
      if (link.linkType === LinkType.NoteId && link.target === targetNoteId) {
        isMatch = true;
      } else if (link.linkType === LinkType.NoteTitle && link.target === targetTitle) {
        isMatch = true;
      }
      
      if (isMatch) {
        const context = extractLinkContext(note.content, link);
        
        backlinks.push({
          fromNoteId: note.noteId,
          toNoteId: targetNoteId,
          fromTitle: note.title,
          linkText: link.displayText || link.target,
          context
        });
      }
    }
  }
  
  return backlinks;
}

/**
 * Validate broken links across a collection of notes
 */
export function validateLinks(
  notes: NoteWithContent[],
  progressCallback?: (progress: { current: number; total: number; message?: string }) => void
): Array<{ note: NoteWithContent; brokenLinks: ParsedLink[] }> {
  const noteMap = new Map<EntityId, NoteWithContent>();
  const titleMap = new Map<string, NoteWithContent>();
  
  // Build lookup maps for efficient validation
  for (const note of notes) {
    noteMap.set(note.noteId, note);
    titleMap.set(note.title, note);
  }
  
  const results: Array<{ note: NoteWithContent; brokenLinks: ParsedLink[] }> = [];
  
  for (let i = 0; i < notes.length; i++) {
    const note = notes[i];
    
    if (!note) {
      continue;
    }
    
    if (progressCallback) {
      progressCallback({
        current: i + 1,
        total: notes.length,
        message: `Validating links in: ${note.title}`
      });
    }
    
    if (!note.content) {
      continue;
    }
    
    const links = parseLinks(note.content, note.noteId);
    const brokenLinks: ParsedLink[] = [];
    
    for (const link of links) {
      let isValid = false;
      
      if (link.linkType === LinkType.NoteId) {
        isValid = noteMap.has(link.target);
      } else if (link.linkType === LinkType.NoteTitle) {
        isValid = titleMap.has(link.target);
      }
      
      if (!isValid) {
        brokenLinks.push(link);
      }
    }
    
    if (brokenLinks.length > 0) {
      results.push({ note: note as NoteWithContent, brokenLinks });
    }
  }
  
  return results;
}

/**
 * Replace links in content with new targets
 */
export function replaceLinkTargets(
  content: string,
  replacements: Array<{ oldTarget: string; newTarget: string; linkType?: LinkType }>
): string {
  let result = content;
  
  for (const replacement of replacements) {
    const { oldTarget, newTarget, linkType } = replacement;
    
    if (linkType === LinkType.NoteId || VALIDATION_PATTERNS.NOTE_ID.test(oldTarget)) {
      // Replace note ID links
      const pattern = new RegExp(
        `\\[\\[${escapeRegex(oldTarget)}(\\|[^\\]]+)?\\]\\]`,
        'g'
      );
      result = result.replace(pattern, `[[${newTarget}$1]]`);
    } else {
      // Replace title links
      const pattern = new RegExp(
        `\\[\\[${escapeRegex(oldTarget)}(\\|[^\\]]+)?\\]\\]`,
        'g'
      );
      result = result.replace(pattern, `[[${newTarget}$1]]`);
    }
  }
  
  return result;
}

/**
 * Convert between note ID and title links
 */
export function convertLinkFormat(
  content: string,
  noteMap: Map<EntityId, NoteWithContent>,
  titleMap: Map<string, NoteWithContent>,
  targetFormat: LinkType
): string {
  const links = parseLinks(content);
  let result = content;
  let offset = 0;
  
  for (const link of links) {
    if (link.linkType === targetFormat) {
      continue; // Already in target format
    }
    
    let newTarget: string | null = null;
    
    if (targetFormat === LinkType.NoteId && link.linkType === LinkType.NoteTitle) {
      // Convert title to note ID
      const note = titleMap.get(link.target);
      if (note) {
        newTarget = note.noteId;
      }
    } else if (targetFormat === LinkType.NoteTitle && link.linkType === LinkType.NoteId) {
      // Convert note ID to title
      const note = noteMap.get(link.target);
      if (note) {
        newTarget = note.title;
      }
    }
    
    if (newTarget) {
      const displayText = link.displayText ? `|${link.displayText}` : '';
      const newLink = `[[${newTarget}${displayText}]]`;
      const adjustedStart = link.startPos + offset;
      const adjustedEnd = link.endPos + offset;
      
      result = result.slice(0, adjustedStart) + newLink + result.slice(adjustedEnd);
      offset += newLink.length - (link.endPos - link.startPos);
    }
  }
  
  return result;
}

/**
 * Extract context around a link for preview/reference
 */
export function extractLinkContext(content: string, link: ParsedLink, contextChars = 100): string {
  const start = Math.max(0, link.startPos - contextChars);
  const end = Math.min(content.length, link.endPos + contextChars);
  
  let context = content.slice(start, end);
  
  // Clean up context boundaries
  if (start > 0) {
    const spaceIndex = context.indexOf(' ');
    if (spaceIndex > 0 && spaceIndex < 20) {
      context = '...' + context.slice(spaceIndex);
    }
  }
  
  if (end < content.length) {
    const lastSpaceIndex = context.lastIndexOf(' ');
    if (lastSpaceIndex > context.length - 20 && lastSpaceIndex > 0) {
      context = context.slice(0, lastSpaceIndex) + '...';
    }
  }
  
  return context.trim();
}

/**
 * Get link statistics for a collection of notes
 */
export function getLinkStatistics(notes: NoteWithContent[]): {
  totalLinks: number;
  uniqueTargets: number;
  brokenLinks: number;
  linkTypes: Record<LinkType, number>;
  mostLinkedNotes: Array<{ noteId: EntityId; title: string; backlinks: number }>;
} {
  const linkCounts = new Map<string, number>();
  const noteMap = new Map<EntityId, NoteWithContent>();
  const titleMap = new Map<string, NoteWithContent>();
  const linkTypes = { [LinkType.NoteId]: 0, [LinkType.NoteTitle]: 0 };
  
  let totalLinks = 0;
  let brokenLinks = 0;
  
  // Build lookup maps
  for (const note of notes) {
    noteMap.set(note.noteId, note);
    titleMap.set(note.title, note);
  }
  
  // Count links and validate
  for (const note of notes) {
    if (!note.content) continue;
    
    const links = parseLinks(note.content, note.noteId);
    totalLinks += links.length;
    
    for (const link of links) {
      linkTypes[link.linkType]++;
      
      // Count backlinks
      const targetKey = link.linkType === LinkType.NoteId ? link.target : link.target;
      linkCounts.set(targetKey, (linkCounts.get(targetKey) || 0) + 1);
      
      // Check if broken
      let isValid = false;
      if (link.linkType === LinkType.NoteId) {
        isValid = noteMap.has(link.target);
      } else {
        isValid = titleMap.has(link.target);
      }
      
      if (!isValid) {
        brokenLinks++;
      }
    }
  }
  
  // Get most linked notes
  const mostLinkedNotes = Array.from(linkCounts.entries())
    .sort((a, b) => b[1] - a[1])
    .slice(0, 10)
    .map(([target, count]) => {
      const note = noteMap.get(target) || titleMap.get(target);
      return {
        noteId: note?.noteId || target,
        title: note?.title || target,
        backlinks: count
      };
    });
  
  return {
    totalLinks,
    uniqueTargets: linkCounts.size,
    brokenLinks,
    linkTypes,
    mostLinkedNotes
  };
}

// ========== Helper Functions ==========

/**
 * Validate link target based on type
 */
function validateLinkTarget(target: string, linkType: LinkType): boolean {
  if (!target || target.length === 0) {
    return false;
  }
  
  if (linkType === LinkType.NoteId) {
    return VALIDATION_PATTERNS.NOTE_ID.test(target);
  } else {
    return VALIDATION_PATTERNS.NOTE_TITLE.test(target) && !target.includes(']]');
  }
}

/**
 * Validate display text
 */
function validateDisplayText(displayText: string): boolean {
  if (!displayText) return true;
  return !VALIDATION_PATTERNS.INVALID_DISPLAY.test(displayText) && displayText.length <= 200;
}

/**
 * Get line offsets for efficient line number calculation
 */
function getLineOffsets(content: string, cacheKey?: EntityId): number[] {
  const now = Date.now();
  
  if (cacheKey) {
    const cached = lineOffsetCaches.get(cacheKey);
    if (cached && cached.content === content && now - cached.lastUpdate < CACHE_CONFIG.CACHE_TTL_MS) {
      return cached.offsets;
    }
  }
  
  const offsets = [0];
  for (let i = 0; i < content.length; i++) {
    if (content[i] === '\n') {
      offsets.push(i + 1);
    }
  }
  
  if (cacheKey) {
    // Manage cache size
    if (lineOffsetCaches.size >= CACHE_CONFIG.MAX_CACHE_SIZE) {
      const oldestKey = lineOffsetCaches.keys().next().value;
      if (oldestKey) {
        lineOffsetCaches.delete(oldestKey);
      }
    }
    
    lineOffsetCaches.set(cacheKey, {
      offsets,
      lastUpdate: now,
      content
    });
  }
  
  return offsets;
}

/**
 * Get line and column number from character position
 */
function getLineAndColumn(position: number, lineOffsets: number[]): { lineNumber: number; columnNumber: number } {
  let lineNumber = 1;
  
  for (let i = 1; i < lineOffsets.length; i++) {
    const offset = lineOffsets[i];
    if (offset !== undefined && position < offset) {
      break;
    }
    lineNumber = i + 1;
  }
  
  const baseOffset = lineOffsets[lineNumber - 1] ?? 0;
  const columnNumber = position - baseOffset + 1;
  
  return { lineNumber, columnNumber };
}

/**
 * Escape regex special characters
 */
function escapeRegex(text: string): string {
  return text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Cleanup expired caches (called periodically)
 */
function cleanupCaches(): void {
  const now = Date.now();
  const expiredKeys: string[] = [];
  
  for (const [key, cache] of lineOffsetCaches.entries()) {
    if (now - cache.lastUpdate > CACHE_CONFIG.CACHE_TTL_MS) {
      expiredKeys.push(key);
    }
  }
  
  for (const key of expiredKeys) {
    lineOffsetCaches.delete(key);
  }
}

// Setup periodic cache cleanup
if (typeof setInterval !== 'undefined') {
  setInterval(cleanupCaches, CACHE_CONFIG.CLEANUP_INTERVAL_MS);
}